Add Docker development environment and install script for A25

MarcOlivierFecteau opened the pull request:
This commit completes the Dockerfile started in A24 for installation pipeline validation, and updates the target to ROS2-Jazzy / Ubuntu 24.04 (in order to be compatible with the RPi5 for A25). The Docker image has been built and run successfully on Windows 11 (23H2) on amd64 architecture, and the baseline codebase (including ROS package dependencies) has been built succesfully inside the container. Note that the "test" program (`ros2 launch racecar_bringup teleop.launch.py`) has **NOT** been tested. Therefore, the baseline codebase **SHOULD** be tested with the proper hardware setup (i.e. racecar with RPi5 w/ Ubuntu24.04) **BEFORE** accepting this PR.

This commit also adds utility scripts for building and running the Docker image on Windows, for ROS environment setup on entering the Docker container, and for building the ROS packages with `colcon`.

NOTES:

- The permissions for accessing the joystick controller and the TTY devices have **NOT** been given, and **SHOULD** be given **BEFORE** testing the baseline codebase (the devices MUST be plugged in in order to add permissions);
- The Dockerfile **COULD** (probably) be optimized to reduce the size of the image.

Commit:
Add Docker development environment for A25

This commit completes the Dockerfile started in A24 for installation
pipeline validation, and updates the target to ROS2-Jazzy / Ubuntu
24.04 (in order to be compatible with the RPi5 for A25). The Docker
image has been built and run successfully on Windows 11 (23H2) on amd64
architecture, and the baseline codebase (including ROS package
dependencies) has been built succesfully inside the container. Note that
the "test" program (`ros2 launch racecar_bringup teleop.launch.py`) has
NOT been tested. Therefore, the baseline codebase SHOULD be tested with
the proper hardware setup (i.e. racecar with RPi5 w/ Ubuntu24.04) BEFORE
accepting this PR.

This commit also adds utility scripts for building and running the Docker
image on Windows, for ROS environment setup on entering the Docker container,
and for building the ROS packages with `colcon`.

NOTES:

- The permissions for accessing the joystick controller and the TTY
  devices have NOT been given, and SHOULD be given BEFORE testing the
  baseline codebase;
- The Dockerfile COULD (probably) be optimized to reduce the size of the
  image.

MarcOlivierFecteau commented:
I would gladly test the baseline codebase on the hardware if you could provide me with a (known-good) racecar. I can provide my own RPi5 (and Arduino Mega2560), but it has 8GB of RAM, and I think the racecar SHOULD be able to run on a RPi5 with 4GB RAM, so if you (SherbyRobotics, or the "Robotics department" of UdeS) could provide me with a 4GB model, it would be ideal.

Commit:
Split ROS installation from Docker image

The requirements for this feature state that the install script
SHOULD be the same for both the Docker image and the direct installation
on a RaspberryPi 5. Therefore, a more 'barebone' base Docker image is
used, and the installation of the package dependencies was moved into
the `ros_entrypoint.sh`.

NOTES:

- The install script was tested inside a Docker container running the
  `Ubuntu 24.04 amd64` base image from Windows 11 amd64. The
  install script SHOULD be tested (and validated) on a RaspberryPi 5 with a
  fresh install of Ubuntu 24.04 arm64 BEFORE merging the PR.
- I tested the install script inside a Docker container running the
  `Ubuntu 24.04 arm64` bsae image from Windows 11 amd64, but I
  encountered the following issues:
  - The install script took more than 1 hour to run;
  - There were errors when trying to find certain packages or files.

Commit:
Minor fixes to install script

This commit adds (temporary) Band-Aids to make the install script not crash:

- `rosdep init` error due to existing list file;
- hard-coded username for chown in `/` workaround;
- add `--cmake-clean-cache` to `colcon build` and links it to `source` to
  avoid skipping build;

The install script has been tested on an RPi5 8GB with an Ethernet
connection. The install script took ~20 minutes to complete. However, the
"test" ROS2 program (`racecar_teleop/teleop.launch.py`) is yet to be tested
on a racecar.

NOTES:

- The install script can easily take 30+ minutes to complete on a RPi5 4GB
with no heatsink/fan;
- The "temporary" Band-Aids SHOULD be replaced before A25 for a more elegant
  solution.

chameau5050 committed:
add devcontainer integration

chameau5050 commented on run.bat:
This file is not needed since we are using the dev container

chameau5050 commented on build.bat:
Convert this script into a linux bash script

chameau5050 commented on Dockerfile:
I have remove the entrypoint to package all dependecy in the image. That way when using docker pull it will be ready to be use

MarcOlivierFecteau replied:
Understood.

chameau5050 reviewed (commented):
Good job! Only a couple of changes needed. Can you also update the documentation on the installation process?

chameau5050 commented on images/setup_vm_ubuntu2404_jazzy.bash:
```diff
@@ -0,0 +1,71 @@
+fi
+rosdep update --rosdistro=${ROS_DISTRO}
+rosdep install --rosdistro=${ROS_DISTRO} --from-paths src --ignore-src -y
+colcon build --cmake-clean-cache && \
+source ${ROS2_DIR}/install/local_setup.bash
```
I have tested the install script on an RPI5 with 4GB of RAM and it appears that we are exceeding the memory limits. I suggest adding the following lines to ensure the build works even on platforms with very constrained memory.

To make the script smart and only slow down on constrained platforms, we could add a condition to check the total memory of the platform. To get the memory, you can use something like [that](https://www.baeldung.com/linux/total-physical-memory) :

``` bash
total_ram () {
  local totalram=$(cat /proc/meminfo | grep -i 'memtotal' | grep -o '[[:digit:]]*')
  echo $totalram
}
ram_size=$(total_ram)
```


```suggestion
export MAKEFLAGS="-j1"
colcon build --cmake-clean-cache --parallel-workers 1 && \
source ${ROS2_DIR}/install/local_setup.bash
```

chameau5050 replied:
Also, while I am thinking about it, I would suggest adding the `--symlink-install` flag when building since the workspace will be used for development.

MarcOlivierFecteau replied:
Suggestions have been added in commit [#90514a5](https://github.com/MarcOlivierFecteau/racecar-a25/commit/90514a5abda8b42c1edb193de56ea90731e3a73d).

Commit:
Add missing tweaks from #3240466

This commits applies the suggestions from chameau5050's review of PR #10.

Changes:

- Removes unused `run.bat` file;
- Replaces Batch Docker build script for its Bash equivalent;
- Formats `devcontainer.json` and `Dockerfile`;
- Adds `-eux` options to install script for "robustness".

Commit:
Update documentation and install script for #10

This commit removes a bunch of outdated (or sub-optimal/inadequate)
documentation, and replaces absolute links in `README`s for relative links
(as it SHOULD be), and formats Markdown files to respect standard formatting
rules. This commit also changes the install script's extension from Bash to
Shell (for "best practices" reasons)

Changes:

- Various updates to READMEs;
- Adds condition to cloning the racecar repo in install script;
- Removes `-u` option in install script due to various conflicts with
environment variables being set outside the script.

Commit:
Add suggestions from chameau5050 to install script

This commits adds a function to get the RAM size of the device (later modi-
fications to the install script will adjust processing resources based on
the device's RAM size to ensure the installation works even on memory
contrained devices). This commit also adds arguments to `colcon build`
command.

NOTES:

- This commit does NOT make use of the added function in the install script.

chameau5050 committed:
rename setup script to bash

chameau5050 committed:
revert documentation

chameau5050 committed:
revert documentation
